src/browser-polyfill.js

```diff
@@ -361,18 +361,44 @@ if (typeof browser === "undefined") {
        *        yield a response. False otherwise.
        */
       return function onMessage(message, sender, sendResponse) {
-        let result = listener(message, sender);
+        let didCallSendResponse = false;
+
+        let wrappedSendResponse;
+        let sendResponsePromise = new Promise(resolve => {
+          wrappedSendResponse = function(response) {
+            didCallSendResponse = true;
+            resolve(response);
+          };
+        });
+
+        let result = listener(message, sender, wrappedSendResponse);
+
+        const isResultThenable = result !== true && isThenable(result);
 
-        if (isThenable(result)) {
+        // If the listener didn't returned true or a Promise, or called
+        // wrappedSendResponse synchronously, we can exit earlier
+        // because there will be no response sent from this listener.
+        if (result !== true && !isResultThenable && !didCallSendResponse) {
+          return false;
+        }
+
+        // If the listener returned a Promise, send the resolved value as a
+        // result, otherwise wait the promise related to the wrappedSendResponse
+        // callback to resolve and send it as a response.
+        if (isResultThenable) {
           result.then(sendResponse, error => {
             console.error(error);
             sendResponse(error);
-          });
-
-          return true;
-        } else if (result !== undefined) {
-          sendResponse(result);
+          }).catch(err => console.error);
+        } else {
```

Comment:
There is another significant, likely wrong change.
Now `sendResponse` is called unconditionally. This causes issues when there are multiple recipients to a message, where only one is expected to send a message.

The logic should be:
- If sendResponse was called before returning, send the response.
- Otherwise if the return value is true, wait for sendResponse.
- Otherwise if the return value is a promise, handle it (call sendResponse upon promise resolution).
- Otherwise, don't call sendResponse. You can either return the value as-is, or not return anything (undefined).

Reply:
@Rob--W There is a `return false` at line 382 which is meant to tell to Chrome that the listener is not going to send any response when the result is not true, is not a thenable and the wrappedSendResponse callback has not already been called by the listener.

After that `return false` the wrapped listener is supposed to send a response, from the returned promise if any, otherwise it means that sendResponse has been called synchronously or the return value is true.

Is there is any additional test case that I may add to the smoke test that is running on a real chrome extension to catch any other scenario that may be missing from your point of view?
(besides the one related to the "rejected promise" scenarios, that at this point it seems better to move in a follow up given that the issue was already there, even without the changes added here for the sendResponse callback support).

Reply:
Oops, I stand corrected.

There is no test coverage for scenarios with more than 1 recipient. This can be tested either via an `all_frames` content script in a tab, or with another extension frame in the background page.

Scenarios:

- Neither replies.
- One replies, the other does not.
- Both reply (maybe with the same value, since the order of receiving messages is unspecified?).

For full coverage, try to run the test twice, once with sendResponse and again with `Promise`.